### DISCUSSION TITLE
Reverse Geocoding

### DISCUSSION THREAD

**Discussion by ugoi** [open-meteo/open-meteo#698] [Ideas]
Can you add a reverse geocoding feature to your geocoding api?
For example like this

https://developers.google.com/maps/documentation/javascript/examples/geocoding-reverse

**Comment by ugoi** [comment]
Your current geocoding api 

https://open-meteo.com/en/docs/geocoding-api

Doesn't support reverse geocoding or does it? If this feature already exists please let me know.

**Reply by patrick-zippenfenig** [reply]
Hi, a reverse geocoding API is on the task list: https://github.com/open-meteo/geocoding-api/issues/6. Unfortunately, there is no ETA and it might take a while until this features is available.

**Comment by ugoi** [comment]
Im Building a weather app and I need to display the location and weather. But if I use the weather api and provide it with coordinates it will only give me the weather of these coordinates but not the city name and country.

**Comment by ugoi** [comment]
And I don't want to use google maps reverse geocoding only to get the city name of the coordinates cause it requires an api key and i need a google account. It's too overkill in my opinion to use google maps api for this.

**Comment by marty-sullivan** [comment]
@ugoi while the work to use it is not quite as simple as an API, the data for reverse geocoding is available for free in Overture's open map dataset. 

https://registry.opendata.aws/overture/ 
https://docs.overturemaps.org/

**Comment by guidocioni** [comment]
You can also use reverse geocoding from mapbox. It does require an account and apikey but you'll never finish the credits of the free tier...

There are really few services giving you reverse geocoding functionality for free, as it is not an easy task to implement.

**Comment by fernandolucchesi** [comment]
Found this one, no API key required :)
https://nominatim.org/release-docs/latest/api/Reverse/

```
interface NominatimResponse {
  address?: {
    city?: string
    town?: string
    village?: string
    municipality?: string
    state?: string
    region?: string
    country?: string
    country_code?: string
  }
  display_name?: string
}

const NOMINATIM_API_URL = 'https://nominatim.openstreetmap.org/reverse'

/**
 * Reverse geocode coordinates to location name (uses Nominatim)
 * @param lat - Latitude
 * @param lon - Longitude
 */
export async function reverseGeocode(
  lat: number,
  lon: number,
): Promise<string | null> {
  const url = new URL(NOMINATIM_API_URL)
  url.searchParams.set('lat', lat.toString())
  url.searchParams.set('lon', lon.toString())
  url.searchParams.set('format', 'json')
  url.searchParams.set('addressdetails', '1')
  url.searchParams.set('accept-language', 'en')

  let response: Response
  try {
    response = await fetch(url.toString(), {
      headers: {
        'User-Agent': 'My App',
      },
    })
  } catch {
    return null
  }

  if (!response.ok) {
    if (response.status === 429) {
      throw new Error('RATE_LIMITED')
    }
    return null
  }

  let data: NominatimResponse
  try {
    data = await response.json()
  } catch {
    return null
  }

  if (data.address) {
    const address = data.address
    const city =
      address.city ||
      address.town ||
      address.village ||
      address.municipality ||
      ''
    const state = address.state || address.region || ''
    const country = address.country || ''

    const locationParts = [city, state, country].filter(Boolean)
    if (locationParts.length > 0) {
      return locationParts.join(', ')
    }
  }

  if (data.display_name) {
    const parts = data.display_name.split(',')
    if (parts.length >= 2) {
      return `${parts[0].trim()}, ${parts[parts.length - 1].trim()}`
    }
    return data.display_name
  }

  return null
}

```
